interpreter/interpretable.go

```diff
@@ -953,6 +991,24 @@ func (e *evalWatchConst) Cost() (min, max int64) {
 	return estimateCost(e.InterpretableConst)
 }
 
+// evalWatchCall describes a watcher of an instCall Interpretable.
+type evalWatchCall struct {
+	InterpretableCall
+	observer evalObserver
```

Comment:
This case should already be handled by `evalWatch`; however, it raises an interesting point.

I think it would be best to update the `evalObserver` interface with a third argument which may be either a `Qualifier` or `Interpretable` value:

```
func evalObserver(id int64, programStep interface{}, value ref.Val)
```

This would allow you to pass the contextual information needed to an observer implementation which could be used to update size without necessarily having to use the state tracking. At least not in the way it's currently done. 

You could in theory maintain a stack of values:

1. `InterpretableCall` values consume argument values off of the value stack, and push one in return.
2. `InterpretableConst` values push an argument value onto the value stack.
3. `Qualifier` values consume an argument value, and push one in return.

There is a missing instruction type, let's call it `InterpretableConstructor` which returns the type of what's being constructed and a list of all `Interpretable` nodes used to init the instance. 

If you have this information, then by the end of the computation you should have a value stack of size 1 which contains the actual cost. It should be slightly better because it allows you to decouple for the state tracking that's currently used, and also remove many of the `Cost` methods on the intepretable nodes since they can be modeled directly within the `Decorator` implementation.

Reply:
Thanks for the suggestion. It makes sense in terms of decoupling the state tracking. I am wondering if it is possible to involve this change into my next PR and get the current one merged to unblock the changes from k/k side first?

Reply:
I'll prototype this real quick and report back.

Reply:
I've committed a change to this PR to address this one.  

I've mostly followed @TristonianJones suggestions, augmented slightly with one of my own goals of being able to have multiple observers on `Interpretable` nodes using only the eval watch decorators. That helped set things up so that any combination of `OptTrackState`, `OptExhaustiveEval` and `OptTrackCost` can be used safely and with minimal overhead.

I added `InterpretableConstructor` like suggested and also added `InterpretableBooleanBinaryOp` to make it ergonomic to to handle `&&` and `||`.

Other than those changes, this is almost an entirely a simplification and consolidation of its logic to a single observer.

Reply:
Thanks @jpbetz  for handling this. I have added a fix to avoid potential panic. It should be ready for the next round of review. @TristonianJones Would you mind to give it a second pass when you have time? Thank you